contracts/contract_call.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// QueryCallFromVM creates a contract Call that peforms a Query using the VMAddress as from
+func QueryCallFromVM(to common.Address, maxGas uint64, msg Message) Call {
+	return &contractCall{
+		readOnly: true,
+		from:     VMAddress,
+		to:       to,
+		maxGas:   maxGas,
+		msg:      msg,
+	}
+}
+
+// WriteCallFromVM creates a contract Call that perfoms a Write operation using the VMAddress as from
+func WriteCallFromVM(to common.Address, maxGas uint64, value *big.Int, msg Message) Call {
+	return &contractCall{
+		readOnly: false,
+		from:     VMAddress,
+		to:       to,
+		maxGas:   maxGas,
+		msg:      msg,
+		value:    value,
+	}
+}
+
+// QueryCallOnRegisteredContract creates a contract Call that performs a Query on a contract from the registry
+func QueryCallOnRegisteredContract(to common.Hash, maxGas uint64, msg Message) Call {
+	return &registeredContractCall{
+		registryId: to,
+		callTemplate: contractCall{
+			readOnly: true,
+			from:     VMAddress,
+			maxGas:   maxGas,
+			msg:      msg,
+		},
+	}
+}
+
+// WriteCallOnRegisteredContract creates a contract Call that performs a Write operation on a contract from the registry
+func WriteCallOnRegisteredContract(to common.Hash, maxGas uint64, value *big.Int, msg Message) Call {
+	return &registeredContractCall{
+		registryId: to,
+		callTemplate: contractCall{
+			readOnly: false,
+			from:     VMAddress,
+			maxGas:   maxGas,
+			value:    value,
+			msg:      msg,
+		},
+	}
+}
+
+// Message represents a msg to a contract
+type Message struct {
+	abi    *abi.ABI
+	method string
+	args   []interface{}
+}
+
+// NewMessage creates a new contract message
+func NewMessage(abi *abi.ABI, method string, args ...interface{}) Message {
+	return Message{
+		abi:    abi,
+		method: method,
+		args:   args,
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Message) encodeCall() ([]byte, error) { return am.abi.Pack(am.method, am.args...) }
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Message) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
+
+// registeredContractCall represents a Call to a contract in the registry
+type registeredContractCall struct {
+	registryId   common.Hash
+	callTemplate contractCall
+}
+
+// Run will execute the call using the caller.
+// Execution result will be populated into result
+func (rc *registeredContractCall) Run(caller vm.EVMCaller, result interface{}) (uint64, error) {
+	contractAddress, err := resolveAddressForCall(caller, rc.registryId, rc.callTemplate.msg.method)
+	if err != nil {
+		return 0, err
+	}
+
+	// copy call template
+	call := rc.callTemplate
+	call.to = contractAddress
+
+	return call.Run(caller, result)
+}
+
+// contractCall represents a Call to a contract
+type contractCall struct {
```

Comment:
I don't think that the introduction of `contractCall` is helping much. It seems to be a somewhat arbitrary abstraction.

If we look at the usage of `Call.Run` vs what was previously ocurring we see:
```go
	_, err := StaticCallFromSystem(evm, params.RegistrySmartContractAddress, getAddressForFuncABI, "getAddressFor", []interface{}{registryId}, &contractAddress, params.MaxGasForGetAddressFor) // old
	_, err := QueryCallFromVM(params.RegistrySmartContractAddress, params.MaxGasForGetAddressFor, NewMessage(&getAddressForFuncABI, "getAddressFor", registryId)).Run(caller, &contractAddress) // new
```

The user needs to provide the same parameters but they now need separate them between 2 function calls, the constructor and Run.

This would potentially be helpful if `contractCall` instances needed to be constructed in a separate location to where run was called or if `contractCall` instances needed to be reused. But in all cases in this PR `contractCall` instances are constructed at the same time as they are used and then discarded, like the code shown above. If all instances of `contractCall` are used immediately after construction then we don't derive any benefit from the introduction of a struct. We may as well just call a function with all the parameters as we did before.

In order to maximise the value of a struct we should try to cluster its fields such that they are likely to change together and such that a single instance of a struct can be reused.

Looking at the fields of `contractCall` we can see that they do not change together.
```go
	readOnly bool // changes with the method and abi which change on redeployments.
	from     common.Address // Always the same, the VMAddress (could be dropped entirely) 
	to       common.Address // Since this is a contract call this changes only with redeployments.
	msg      Message  // (abi, method and args) abi, and method change with redeployments, args will likely change on every call.
	value    *big.Int // Likely to change on every call.
	maxGas   uint64 // A value hardcoded in the code, could change on redeployments or code changes.
```

If we instead clustered the fields of this struct such that all struct fields are likely to change in response to redeployments or code changes then we would be able to construct a struct at node startup and reuse it for the lifetime of a node.

Such a struct would have the following fields, with the args and value being provided at runtime.
```go
        readOnly
	address common.Address
	method  string
	abi     *abi.ABI
	maxGas  uint64
```

I've had a go at this [here](https://github.com/celo-org/celo-blockchain/compare/731425bc2f935729d1bf2d0ce0f88a4b053d1292..piersy/evm-refactor-2) its also a bit leaner than its predecessor, dropping ~30 LOC, but more importantly setting up a framework where we can reuse instances for the lifetime of a node.

Reply:
For someone that has been doing golang for so many years, you surely give # of LOC a lot of importance! Try haskell! generics, and all that stuff. Golang doesn't very well suited to give you a lot of expressive power in a short number of LOC! So i don't really measure it by LOC, but readability (which of course is subjective)

I like the argument of being able to reuse the struct and i like the `ContractMethod` struct. I'm not very fond of ending with a 9 positional arguments `call()` after that. That's what i was trying to avoid in the first place. Quite a well known [code smell](https://sourcemaking.com/refactoring/smells/long-parameter-list). 

On other languages, to solve it, you can try to split the function in parts, so to avoid doing ALL in one place.  Have smaller function, encode different parts in different functions. That's why i created the `Message` struct, to try to encode the abi part of it, and also because using "vargs" as opposed to a `interface{}[]` seemed more natural to me, and provides a better UX (at least in my opinion, again ux can be subjective).

Another strategy to reduce the parameter bloat, if to have property objects. You see a lot of that in javascript/typescript. Or if the language supports it "named arguments". The `ContractCall` object it's a parameter object in disguise; it provides names for all the things you need to pass, but  instead of passing a Call object to a function,  you create a call and then call `Run()`.

That also has a secondary benefit. I'm able to add new behaviour in "types of call" without modyfing the interface; nor adding conditionals to the code (and longer functions). That's ["replacing conditional for polymorphism"](https://refactoring.guru/replace-conditional-with-polymorphism); so let's the caller worry less how it works, and just do a call.

Finally, the idea of have a reusable call did cross my mind. But i didn't find much reusable parameters. Surely `(abi, method, maxGas)` seems to be reusable. `ContractAddress` comes from the registry is can change. Even `maxGas` is an argument for some calls (check `getBalance`). So i didn't end up with nothing too convincing.

In any case, even if no reuse, creating a object just to make a call and consume it; it's nothing an unusual pattern in my experience. Actually `Call` interface can be seen as an implementation of [Command design pattern](https://www.gofpatterns.com/behavioral/patterns/command-pattern.php) from the classic GOF book.

Reply:
I've tried to get the "Best of both worlds" and create something that can be reused. Check this commit https://github.com/celo-org/celo-blockchain/pull/1513/commits/da5f69432532dadefeef92738ed53476320396e1

Basically i've tranformed `Message` into `Method` which contains (abi, method, maxGas)

And then created `BoundedMethod` which also contains the contract address (in fact the ability to resolve the address).

The benefit from this approach is that now i can initialize at a module level `BoundMethods` which can greatly simply the code of calling them.

For example:
```
var getWhitelistMethod = contracts.NewRegistryContractMethod(params.FeeCurrencyWhitelistRegistryId, &getWhitelistFuncABI, "getWhitelist", params.MaxGasForGetWhiteList)

func CurrencyWhitelist(caller vm.EVMCaller) ([]common.Address, error) {
	returnList := []common.Address{}
	_, err := getWhitelistMethod.VMQuery(caller, &returnList)
       return returnList, err
}
```


Check the demo here https://github.com/celo-org/celo-blockchain/pull/1522

Reply:
Hey @mcortesi Sure go is verbose, but when I consider LOC its not really the absolute value of LOC I'm interested in, its the change. Did LOCs go up? If so the code likely got more complex, and conversely did they go down, its likely that things got simpler. Its a language agnostic measure.

You are right about `call`, it had a lot of parameters, but it was also a private function of that package, so users of the package would not need to interact with it and it was being called in only 2 places. So although I generally prefer less parameters, if usage of functions with large numbers of parameters is contained within a small space I don't see it as a big problem.

I could have made it a function on `ContractCaller` which would save 4 params, but given that I'd already written the code calling `call` it wouldn't have helped me, and I don't think that readability would be changed by doing that (If anything I think readability would be slightly degraded by converting it to a method because the reader would then have to refer back to the struct definition to know what the types of the fields are vs just looking at the parameter definition).

I don't think how the `call` function is presented (as a method or as a pure function) really has much of an impact on the code, I wouldn't have a preference either way. The only thing that would push me one way or the other would be these questions:

* Do we need the functionality of `call` in another struct? If yes then we can make it a pure function and share it between structs, or duplicate the logic between structs, I would be in favour of having a pure function.
* Do we make lots of calls to `call` within a single struct? In this case it looks like it makes sense to make it a method and save repeatedly passing the parameters.

In my branch it was too early to say which of the above situations would be the case.

Sometimes lots of parameters is a code smell but I would generally try to asses this on a case by case basis, using LOC to guide me to an optimal solution, which may end up having a function with lots of params.

Would you consider the new `run` function to still exhibit a code smell? According to the [article](https://sourcemaking.com/refactoring/smells/long-parameter-list) you linked it still exhibits a code smell because it has:
> More than three or four parameters

It has 6. I don't consider it a code smell, but I didn't consider it a code smell with 9. The fact is you need 5 parameters just to call the EVM, does that make all of our code that calls the EVM bad? I don't think so. It could be argued that the EVM is badly designed and that we should refactor it, but sometimes you are doing something complicated and you just need lots of parameters. In those cases I think large numbers of parameters can be justified, there are of course cases where functions can be nicely refactored to remove the large parameter requirements and its good for the code, but I would be wary of trying to force the parameter count down by any means necessary.

With regards to parameter objects sometimes they can help, but if they are being used to just put stuff inside and then get it back out its not really helping, I would say they are hiding the problem.

E.G.

```go
// Callsite
printAll("a", "b", "c", "d", "e", "f")

// Implementation
func printAll(s1, s2, s3, s4, s5, s6) {
    println(s1)
    println(s2)
....
}


```

So we can fix this with a parameter object, you can probably guess :) but just for the purposes of comparison.

```go

// Callsite
printAll(ss{"a", "b", "c", "d", "e", "f"})

// Implementation
func printAll(ss params) {
    println(params.s1)
    println(params.s2)
....
}
```

This solves the "code smell" we are down to one parameter, but the reality is that the function became a little bit harder to call and a little bit more complex in the implementation and we now have the additional code for the struct which is not shown.

The `ContractCall` implementation as it was originally, was like this parameter object example above. It just packed up a bunch of parameters together and then the `Run` method needed to unpack them.

Where previously the caller would write:
```go
return contract_comm.MakeStaticCall(registryId, abi, method, args, returnObj, gas, header, state)
```
The introduction of `ContractCall` means the caller would have to write this:
```go
caller, err := getCaller(header, state)
if err != nil {
	return 0, err
}

return contracts.QueryCallOnRegisteredContract(registryId, gas, contracts.NewMessage(&abi, method, args...)).Run(caller, returnObj)
```

Even if we discount the code required to get the caller, I think we could agree that it's clearly not easier than what went before, its just different. Parameters have been bundled up in a different way, technically it reduced a "code smell", but actually the underlying problem of the code smell remains, which is that the caller when making a contract call is faced with collating a large bunch of parameters, and that is a complex operation.

What's not shown above is the cost of introducing the `ContractCall` which is about 100 LOC, and this in general does increase complexity, its more lines for people to read and understand before they can use the code, its more places that bugs can hide.

If the construction of the `ContractCall` and `Message` objects were to happen elsewhere such that callers only need to call `Run` then that would likely be an improvement because it would have really reduced the burden on the caller to collate so many parameters at the point of making a call. But because the contract call object includes the `args` its seems unlikely to me that you would be able to construct the contract call elsewhere because args are likely to be unique for each call.

You mention ["replacing conditional for polymorphism"](https://refactoring.guru/replace-conditional-with-polymorphism) I'd agree that your design makes that easier, but there aren't any conditionals currently, and its not clear that there would be, so I would consider introducing `ContractCall` now for this reason would be an instance of premature abstraction.

You also mention that Call interface can be seen as an implementation of the Command design pattern. But if you look at the description in the [link](https://www.gofpatterns.com/behavioral/patterns/command-pattern.php) they say.

> The Command pattern encapsulates a request in an object, which enables you to store the command, pass the command to a method, and return the command like any other object.

`ContractCall` was not being passed, stored or returned so from the point of view of GOF it was a needless application of the pattern.

So to make `ContractCall` a useful addition I think it needs to be something that can be re-used/passed and I think this is what your latest changes are doing with `Method` and `BoundMethod`.

Reply:
As i stated on the other comment:

> Did LOCs go up? If so the code likely got more complex, and conversely did they go down, its likely that things got simpler. Its a language agnostic measure.

i don't believe this to be true. But again, i don't have to continue with this "who's right!" kinda of conversation. Too much time is wasted on it; and no real gain.

Just note that on you example, even if you don't reduce the number of parameter, using a parameter object let;s you give name to them, which is why it's help reduce the smell. You won't argue that accessing your parameter by `args[5]` is not as good as saying `args.price`. And when you read a contract call that uses 9 parameters, you'll definietly need some editor tooltip or tool to let you know which parameter is whchi, and you'll have to double check that you're passing the right parameter in the right position. Specially is parameters have the same type.

So, even if it's not good as not needing them, i do believe it makes it clearer.

But again, you seem to like the BoundMethod, so let's leave at that.

Reply:
Hey @mcortesi, I wouldn't say this a a "whos right" conversation, hopefully we get to understand each other better and ideally we come to a shared understanding (not quite there yet :) maybe soon!

Totally agree that `args.price` is preferable to `args[5]` but I'd be interested to know where you would apply this? `args []interface{}` is passed around quite a lot in this PR but its just getting passed through to the vm. 

My IDE actually tells me the name and type of function parameters as I'm typing so for me its equivalent to it auto-completing a parameter object field so I don't find one approach better than the other.

I generally try to reserve the use of parameter objects for times when a large number of parameters may be nil, so rather than filling out a function call with a whole bunch of nils you can just initialise a struct with a couple of fields. I find this good for config objects.

I do like the `BoundMethod` it moves lots of config to a one-off initialisation step, and this definitely reduces the parameter burden when calls :D

Reply:
no worries...! we'll get there

And i agree, `BoundMethod` makes it much simpler by providing this one-off initializatiion step outside. Thanks to your suggestion of having things that can be reused in the struct.

To clarify, on my `args[5]` example. i've used an array as a parallel to positional argument. I mean when you have a function:

```go
func doSomething(a,b,c,d,e,f,g,h uint)
```

in the function implementation you have names (you use `a`,`b,` and not `args[5]`). But on the call site

```
doSomething(1,2,4,5,2,5,8)
```
It's harder to see which is which, and to detect if you have a bug by confusing the parameters. And i personally had to double check that when i see some function calls on the current codebase (on other parts of master). I use vscode tooltip, to double check. But it's not immediately evident. So that's why i prefer it.

The only bad thing about parameter objects is that someone can forget to initialize a field, and go doesn't provide a good way of checking that (unless you create a construction function with positional parameters :( ). So in the future if you add a field, you need to double check all call sites, to see it's constructed properly.

I guess golang doesn't really help. a lot in defensive programming